File: python/173_二叉搜索树迭代器_medium.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: TreeNode):
        self.ans = []
        self.inorder(root)
        self.N = len(self.ans)
        self.idx = 0

    
    def inorder(self, root):
        if not root:
            return
        self.inorder(root.left)
        self.ans.append(root.val)
        self.inorder(root.right)
        return

    def next(self) -> int:
        res = self.ans[self.idx]
        self.idx += 1
        return res

    def hasNext(self) -> bool:
        return self.idx < self.N



# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: cpp/173_二叉搜索树迭代器_medium.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    vector<int> ans;
    int idx;
    void inorder(TreeNode *root, vector<int> &ans) {
        if (!root) return;
        inorder(root->left, ans);
        ans.push_back(root->val);
        inorder(root->right, ans);
        return;
    }
    vector<int> inorderTravelsal(TreeNode *root) {
        vector<int> res;
        inorder(root, res);
        return res;
    }
public:
    BSTIterator(TreeNode* root): idx(0), ans(inorderTravelsal(root)) {}
    
    int next() {
        return ans[idx++];
    }
    
    bool hasNext() {
        return idx < ans.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */